### DISCUSSION TITLE
How to wait for HTTP port listening during app start?

### DISCUSSION THREAD

**Discussion by Rotkaf** [playframework/playframework#13008] [Q&A] [answered]
I haven't found the answer in official docs (https://www.playframework.com/documentation/3.0.x) so I'm asking here:

When my app starts I do some requests for some data to external service:

Simplified extract from code:

```java
public class StartModule extends AbstractModule {

    @Override
    protected void configure() {
        bind(ApplicationStart.class).asEagerSingleton();
	    // other code
	}
}
```	
```java
@Singleton
public class ApplicationStart {

    // private final .... (DI fields)
	
    @Inject
    public ApplicationStart(ApplicationLifecycle lifecycle, Environment environment, MyWsService wsService, MyUtils utils, (...) ) {

        wsService.someRequest().thenAccept(jsonNode -> utils.putDataInCache());

        (...)
    }
}
```

So it gets data and put it into cache for later user for clients.
This request takes unfortunately much time (> 30 seconds) - let's assume we can't improve that.	

The problem:
The listening to HTTP port ("PekkoHttpServer][503][] Listening for HTTP on ...") starts **earlier** than this request got finished.
So when someone invokes getting cached data  by Controller **before** wsService.someRequest() is finished in initialization code it'll call another request to external service because cache is still empty. 
(of course in prod environment it could be many Controller requests during for example one second = many requests to external service...).

I suppose the solution is somewhere near ApplicationLoader (GuiceApplicationBuilder?) but I can't do fully working compliant/working code.
I asked the ChatGPT (and others AI tools) to get some help but the answers were non-compilable or much tricky (for example - set initial port to 0 and after some time delay set to required port number - I don't feel is right elegant solution).

So - is there any recommended solution to the problem?

**Comment by irumiha** [comment]
How are you deploying your service? If you are deploying on Kubernetes you can set up a readiness endpoint that returns a "not ready" status until the cache is populated. K8s will not route traffic to your instance until the readiness probe succeeds.

**Comment by Rotkaf** [comment]
No, it is simply standalone - by running the script in the bin directory (https://www.playframework.com/documentation/3.0.x/Deploying#Using-the-dist-task)

**Comment by ihostage** [comment, marked as answer]
Can you try to block initialization of your `ApplicationStart` component?
```java
wsService.someRequest().thenAccept(jsonNode -> utils.putDataInCache())
  .toCompletableFuture().get();
``` 
It's not a good solution, but can help in your case, because application will wait initialization your component before listen port and handle requests.

PS: Also as an example of this approach you can see [Play Flyway plugin](https://github.com/playframework/flyway-play/tree/main). This plugin does database migration before start an application. It also use eager component with blocked initialization to make all migrations before handle user request.

**Comment by Rotkaf** [comment]
Yes, it works. Thanks for replay and suggestions.
However I'm still little surprised that we have to search for some workarounds for that case (it seemed to me that it is quite common one...).
